File: apps/backend/src/cli/tools/validation.go
// src/cli/tools/validation.go
package tools

import (
	"fmt"

	"github.com/lucas/oraculo/apps/backend/src/db"
	"github.com/lucas/oraculo/apps/backend/src/output"
	"github.com/spf13/cobra"
)

// newValidationCmd returns the "validation" parent command with subcommands.
func newValidationCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "validation",
		Short: "Manage QA validations",
	}
	cmd.AddCommand(newValidationSaveCmd())
	return cmd
}

func newValidationSaveCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "save <story-name>",
		Short: "Save a QA validation verdict for a story",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, _, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			verdictStr, _ := cmd.Flags().GetString("verdict")
			if verdictStr != "approved" && verdictStr != "rejected" {
				err := fmt.Errorf("invalid verdict %q: must be approved or rejected", verdictStr)
				output.WriteError(w, err)
				return err
			}

			findings, _ := cmd.Flags().GetString("findings")

			storyStore := db.NewStoryStore(database)
			story, err := storyStore.GetByName(epicID, name)
			if err != nil {
				output.WriteError(w, err)
				return err
			}

			valStore := db.NewValidationStore(database)
			validation, err := valStore.Save(story.ID, verdictStr, findings)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, validation)
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	cmd.Flags().String("verdict", "", "Validation verdict: approved|rejected (required)")
	_ = cmd.MarkFlagRequired("verdict")
	cmd.Flags().String("findings", "", "QA findings as JSON string")
	return cmd
}

File: apps/orchestrator/src/cli/tools/validation.ts
import { Command } from "commander";
import { EpicStore } from "../../db/stores/epic-store.ts";
import { StoryStore } from "../../db/stores/story-store.ts";
import { ValidationStore } from "../../db/stores/validation-store.ts";
import { writeJSON, writeError } from "../../utils/output.ts";
import { dbFromCmd } from "../context.ts";

export function validationCommands(): Command {
  const cmd = new Command("validation").description("Manage QA validations");

  cmd.addCommand(validationSaveCmd());

  return cmd;
}

function validationSaveCmd(): Command {
  return new Command("save")
    .description("Save a QA validation verdict for a story")
    .argument("<story-name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .requiredOption(
      "--verdict <verdict>",
      "Validation verdict: approved|rejected",
    )
    .option("--findings <findings>", "QA findings as JSON string", "")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const verdict = opts.verdict as string;
        if (verdict !== "approved" && verdict !== "rejected") {
          writeError(
            new Error(
              `invalid verdict "${verdict}": must be approved or rejected`,
            ),
          );
          return;
        }

        const db = dbFromCmd(cmd);
        const epicStore = new EpicStore(db);
        const epic = epicStore.getByName(opts.epic);

        const storyStore = new StoryStore(db);
        const story = storyStore.getByName(epic.id, name);

        const valStore = new ValidationStore(db);
        const validation = valStore.Create(
          story.id,
          verdict,
          opts.findings,
        );
        writeJSON(validation);
      } catch (err) {
        writeError(err);
      }
    });
}
